### PR TITLE
Feature/fancy pointers

### PR DESCRIPTION
**Pull request by LukasKerk**
This generalizes the data structure a bit more to be able to handle Allocators that use fancy pointers.

### CODE REVIEW

File: tests/CMakeLists.txt [hunk 1]
```diff
@@ -19,8 +24,8 @@ endif()
 
 # Boost::unit_test_framework
 find_package(Boost 1.54.0 REQUIRED COMPONENTS unit_test_framework)
-target_link_libraries(tsl_sparse_map_tests PRIVATE Boost::unit_test_framework)
```
Comment (bigerl) [inline tests/CMakeLists.txt:22]: unnecessary whitespace change

File: tests/CMakeLists.txt [hunk 2]
```diff
@@ -19,8 +24,8 @@ endif()
 
 # Boost::unit_test_framework
 find_package(Boost 1.54.0 REQUIRED COMPONENTS unit_test_framework)
-target_link_libraries(tsl_sparse_map_tests PRIVATE Boost::unit_test_framework)   
+target_link_libraries(tsl_sparse_map_tests PRIVATE Boost::unit_test_framework)
 
 # tsl::sparse_map
 add_subdirectory(../ ${CMAKE_CURRENT_BINARY_DIR}/tsl)
-target_link_libraries(tsl_sparse_map_tests PRIVATE tsl::sparse_map)  
+target_link_libraries(tsl_sparse_map_tests PRIVATE tsl::sparse_map)
```
Comment (bigerl) [inline tests/CMakeLists.txt:31]: unnecessary whitespace change

File: include/tsl/sparse_hash.h [hunk 3]
```diff
@@ -1009,15 +1063,15 @@ class sparse_hash : private Allocator,
 
   using key_type = typename KeySelect::key_type;
   using value_type = ValueType;
-  using size_type = std::size_t;
-  using difference_type = std::ptrdiff_t;
   using hasher = Hash;
   using key_equal = KeyEqual;
   using allocator_type = Allocator;
-  using reference = value_type &;
+  using reference = value_type &; //does it need to be changed for fancy pointers?
```
Comment (bigerl) [inline include/tsl/sparse_hash.h]: remove comment?

File: tests/fancy_pointer/sparse_hash_set_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+                                  set_sorted.begin()),
+                       "const iterating over the set didn't work");
+}
+
+template <typename T>
+void erase(std::initializer_list<typename T::value_type> l, typename T::value_type extra_value) {
+    auto set = details::default_construct_set<T>();
+    set.insert(extra_value);
+    set.insert(l.begin(), l.end());
+    //find is not tested yet
```
Comment (bigerl) [inline tests/fancy_pointer/sparse_hash_set_tests.cpp]: todo: test find

File: tests/fancy_pointer/sparse_hash_set_tests.cpp [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+    set.erase(iter);
+    BOOST_TEST_REQUIRE(details::is_equal(set, l), "erase did not work as expected");
+}
+
+template <typename T>
+void erase_with_const_iter(std::initializer_list<typename T::value_type> l, typename T::value_type extra_value) {
+    auto set = details::default_construct_set<T>();
+    set.insert(extra_value);
+    set.insert(l.begin(), l.end());
+    //find is not tested yet
```
Comment (bigerl) [inline tests/fancy_pointer/sparse_hash_set_tests.cpp]: todo: test find

File: include/tsl/sparse_hash.h [hunk 6]
```diff
@@ -1343,25 +1399,28 @@ class sparse_hash : private Allocator,
    */
   iterator begin() noexcept {
     auto begin = m_sparse_buckets_data.begin();
-    while (begin != m_sparse_buckets_data.end() && begin->empty()) {
+    //fancy pointers seem to have a problem with "->" in this context
```
Comment (bigerl) [inline include/tsl/sparse_hash.h]: point out where it is because of std::vector

**Review by bigerl — CHANGES_REQUESTED** [review]
looks fine otherwise

File: include/tsl/sparse_hash.h (hunk 3 above)
Comment (LukasKerk) [reply include/tsl/sparse_hash.h]: it is now removed

File: tests/fancy_pointer/sparse_hash_set_tests.cpp (hunk 4 above)
Comment (LukasKerk) [reply tests/fancy_pointer/sparse_hash_set_tests.cpp]: The comment wasn't updated yet. find is already tested, however the code forces the  non-const version of the iterator so find cannot be used.

File: tests/fancy_pointer/sparse_hash_set_tests.cpp (hunk 5 above)
Comment (LukasKerk) [reply tests/fancy_pointer/sparse_hash_set_tests.cpp]: The comment wasn't updated yet. find is already tested, however the code forces the  const version of the iterator so find cannot be used.

File: include/tsl/sparse_hash.h (hunk 6 above)
Comment (LukasKerk) [reply include/tsl/sparse_hash.h]: Comments have been updated.
